### ISSUE TITLE
Autocomplete doesn't work corrently on certain folder names

### ISSUE THREAD

**Issue by Racle** [ohmyzsh/ohmyzsh#11621]
### Describe the bug

Autocomplete doesn't work on certain foldernames. In my case it usually adds "0-" before my search term and "-" to end.

This has been issue at least for year IIRC. 

I started fresh in root account, installed ohmyzsh (with curl) and got same results as in my personal config.

Quick test seems to be indicating that if first folder has 2 letters before `-` and second folder has 3 letters before `-`, then bug seems to appear. 

I included my test cases and asciinema recording also.

### Steps to reproduce

1. Install ohmyzsh
2. run `mkdir my-longfoldername-firstname  you-longfoldername-secondname`
3. type `cd long<tab><tab>` 

### Expected behavior

It should autocomplete like it would in case of:

1. Install ohmyzsh
2. run `mkdir my-dir-2  your-dir-1`
3. type `cd dir<tab><tab>` 
4. autocomplete works as intended

### Screenshots and recordings

https://asciinema.org/a/qqwzYW445NxNEwaPERNwoPc3a

### OS / Linux distribution

Pop_os 22.04

### Zsh version

5.8.1

### Oh My Zsh version

master (673b9fc)

### Terminal emulator

Kitty 0.28.0

### If using WSL on Windows, which version of WSL

None

**Comment by rwmitchell** [comment]
Where is your _cd defined?  Mine comes from share/zsh/5.8.0.2-dev/functions/_cd and doesn’t exhibit this behavior

**Comment by Racle** [comment]
> Where is your _cd defined? Mine comes from share/zsh/5.8.0.2-dev/functions/_cd and doesn’t exhibit this behavior.
> 
> Your version is newer.

How do I check this out?

This also happens on my server. Ubuntu 22.04, zsh 5.8.1 and master (4c82a2ee)

**Comment by Racle** [comment]
Found one way:

```
# whence -v _cd
_cd is a shell function from /usr/share/zsh/functions/Completion/Zsh/_c 
```

**Comment by carlosala** [comment]
I was able to reproduce using `zsh` docker image with with `5.8.1`, but it's not reproducible anymore when using `5.9`.
I wasn't able to fix that issue using completion, but it seems it's some regression with case insensitive completion. It's something from zsh itself, and it's not only related to `cd`, so we can't do anything. You can workaround that error adding `CASE_SENSITIVE=true` in your `zshrc` before sourcing `oh-my-zsh.sh`. This will make completion to be case sensitive but also avoid this error.
Thanks for the report!

**Comment by Racle** [comment]
@carlosala updated to zsh 5.9, but I still have same issue on local machine.

If I do `mkdir my-longfoldername-firstname  you-longfoldername-secondname` and try `folder<tab><tab>` it works just fine.

But `long<tab><tab>` it  doesn't work (it just displays `cd O-longfoldername-`)

And zsh without ohmyzsh doesn't autocomplete anything :thinking:

**Comment by carlosala** [comment]
which is your output to `echo $ZSH_VERSION`?

**Comment by Racle** [comment]
> which is your output to `echo $ZSH_VERSION`?

5.9

asciinema record with omz version and $ZSH_VERSION: https://asciinema.org/a/yzAu4UzQeFxljhq43p9erGNXE

**Comment by Racle** [comment]
@carlosala 

Dockerfile:
```
FROM ubuntu:23.04
RUN apt-get update && apt-get install -y zsh curl git
RUN sh -c "$(curl -fsSL https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/master/tools/install.sh)"
RUN mkdir my-longfoldername-firstname  you-longfoldername-secondname
CMD zsh
```

Build with `docker build -t zsh-test .`
Then run `docker run -ti zsh-test`

`cd long<tab>` will produce the error while `cd fold<tab><tab>` works.

And I assume that autocomplete comes from ohmyzsh as zsh doesn't do autocomplete from middle of name (or at least stock installation doesn't do that)

**Comment by Racle** [comment]
So this seems to happen with ubuntu based systems 🤔

Maybe some system packet handles something diffrently. 

But very odd indeed.

**Comment by roeniss** [comment]
macOS m1 can't reproduce this issue. As Racle suggested this could be OS issue?

<img width="890" alt="Screenshot 2023-04-20 at 01 32 28 by seonghyeok" src="https://user-images.githubusercontent.com/26613280/233141219-3eac5205-6db6-42bf-b66a-b496cc6c53b7.png">

**Comment by carlosala** [comment]
@Racle if you don't mind, could you pack all this info and send an email to https://zsh.org/mla users mailing list?
Thanks!

**Comment by Racle** [comment]
@carlosala just send email (was my first mailing list email, so I hope this went correctly :sweat_smile:)

**Comment by roeniss** [comment]
sorry for interrupting -- I want to reply on https://zsh.org/mla/users/2023/msg00439.html but not sure how to that (never experienced about mailing list). Can you help me please?

**Comment by mcornella** [comment]
@roeniss you have to subscribe to the mailing list and wait for someone to reply again to the mail thread so you get an email that you can reply on, or if that's not an option send a new email to the mailing list referencing that thread by URL or number. See https://zsh.org/mla

**Comment by piiq** [comment]
> I was able to reproduce using `zsh` docker image with with `5.8.1`, but it's not reproducible anymore when using `5.9`. I wasn't able to fix that issue using completion, but it seems it's some regression with case insensitive completion. It's something from zsh itself, and it's not only related to `cd`, so we can't do anything. You can workaround that error adding `CASE_SENSITIVE=true` in your `zshrc` before sourcing `oh-my-zsh.sh`. This will make completion to be case sensitive but also avoid this error. It seems like zsh team solved it in `5.9`, try to update if you can. Thanks for the report!

Uncommenting the `CASE_SENSITIVE=true` setting in `~/.zshrc` solved the issue for me (macos + latest oh-my-zsh), thanks

The simplest reproducible scenario i could come up with is 
1. Create 2 folders `tmp_te` and `tmp_sq`. In my case it was important that the folders have the same prefix and that one ends with _te and the suffix of the second one starts with _s. This looks like it is voodoo but i hope there is some rational explanation to the fact that it's specifically `_te` not working with `_s`.
2. In that folder try `ls te<tab>` and that would complete `ls tmp_T` instead of `ls tmp_`

The fact that in all the cases that i've encountered this behaviour there was a capital letter involved made me think it does have to do something with case sensitive setting. I've tried. It worked. Thanks a lot for the hint. 👏🙏❤️🚀

**Comment by carlosala** [comment]
Closing this as it's something from Zsh and not from our side!
